org.eclipse.xtext.tests/src/org/eclipse/xtext/xtext/XtextGrammarSerializationTest.java

```diff
@@ -0,0 +1,199 @@
+				"StartRule returns fooMM::T:\n" +
+				"	name=ID;\n";
+		doTestSerialization(model, expectedModel);
+	}
+
+	@Test
+	public void testSerializationWithCardinalityOverride() throws Exception {
+		String model =
+				"grammar foo with org.eclipse.xtext.common.Terminals\n" +
+				"generate mm \"http://bar\" as fooMM\n" +
+				"StartRule returns fooMM::T: (name+=ID?)+;\n";
+		String expectedModel =
+				"grammar foo with org.eclipse.xtext.common.Terminals\n" +
+				"\n" +
+				"generate mm \"http://bar\" as fooMM\n" +
+				"\n" +
+				"StartRule returns fooMM::T:\n" +
+				"	(name+=ID)*;\n";
+		doTestSerialization(model, expectedModel);
+	}
+
+	@Test
+	public void testSerializationSuperCall() throws Exception {
+		String model =
+				"grammar foo with org.eclipse.xtext.common.Terminals\n" +
+				"generate mm \"http://bar\" as fooMM\n" +
+				"StartRule returns fooMM::T: name=super::ID value=Terminals::STRING thing=STRING;\n" +
+				"terminal STRING: super;\n";
+		String expectedModel =
+				"grammar foo with org.eclipse.xtext.common.Terminals\n" +
+				"\n" +
+				"generate mm \"http://bar\" as fooMM\n" +
+				"\n" +
+				"StartRule returns fooMM::T:\n" +
+				"	name=super::ID value=Terminals::STRING thing=STRING;\n" +
+				"\n" +
+				"terminal STRING:\n" +
+				"	super;\n";
+		doTestSerialization(model, expectedModel);
+	}
+
+	private void doTestSerialization(String model, String expectedModel) throws Exception {
+		XtextResource resource = getResourceFromString(model);
+		Assert.assertTrue(resource.getErrors().isEmpty());
+		Grammar g = (Grammar) resource.getParseResult().getRootASTElement();
+		Assert.assertNotNull(g);
+		OutputStream outputStream = new ByteArrayOutputStream();
+		resource.save(outputStream, SaveOptions.newBuilder().format().getOptions().toOptionsMap());
+		String serializedModel = outputStream.toString();
+		Assert.assertEquals(LineDelimiters.toPlatform(expectedModel), serializedModel);
+	}
+
+	@Test
+	public void testMetamodelRefSerialization() throws Exception {
+		String model =
+				"grammar foo with org.eclipse.xtext.common.Terminals\n" +
+				"import \"http://www.eclipse.org/2008/Xtext\" as xtext\n" +
+				"generate mm \"http://bar\" as fooMM\n" +
+				"Foo returns fooMM::Foo: name=ID (nameRefs+=NameRef)*;\n" +
+				"NameRef returns xtext::RuleCall : rule=[xtext::ParserRule];\n" +
+				"MyRule returns xtext::ParserRule : name=ID;\n";
+		String expectedModel =
+				"grammar foo with org.eclipse.xtext.common.Terminals\n" +
+				"\n" +
+				"import \"http://www.eclipse.org/2008/Xtext\" as xtext\n" +
+				"generate mm \"http://bar\" as fooMM\n" +
+				"\n" +
+				"Foo returns fooMM::Foo:\n" +
+				"	name=ID (nameRefs+=NameRef)*;\n" +
+				"\n" +
+				"NameRef returns xtext::RuleCall:\n" +
+				"	rule=[xtext::ParserRule];\n" +
+				"\n" +
+				"MyRule returns xtext::ParserRule:\n" +
+				"	name=ID;\n";
+		doTestSerialization(model, expectedModel);
+	}
+
+	public void _testXtestSerializationSelfTest() throws Exception {
```

Comment:
Is this used?

Reply:
no. i wanted to keep this for whatever it was good for.